python/rascaline/rascaline/utils/splines.py

```diff
@@ -0,0 +1,393 @@
+"""
+Splined radial integrals
+========================
+
+Classes for generating spline points which can be used as input for the for the radial
+integral hyper paramater option. For an complete example for the LE how to use the
+classes see :ref:`example-splines`.
+
+.. autoclass:: rascaline.utils.RadialIntegralSplinerBase
+    :members:
+    :show-inheritance:
+
+.. autoclass:: rascaline.utils.RadialIntegralFromFunction
+    :members:
+    :show-inheritance:
+"""
+
+from abc import ABC, abstractmethod
+from typing import Callable, Dict, Optional, Union
+
+import numpy as np
+
+
+class RadialIntegralSplinerBase(ABC):
+    """Base class for splining arbitrary radial integrals.
+
+    If ``_radial_integral_derivative`` is not implemented in a child class it will
+    computed based on finite differences.
+
+    :parameter max_angular: number of radial components
+    :parameter max_radial: number of angular components
+    :parameter spline_cutoff: cutoff radius for the spline interpolation. This is also
+        the maximal value that can be interpolated.
+    :parameter accuracy: accuracy of the numerical integration and the splining.
+        Accuracy is reached when either the mean absolute error or the mean relative
+        error gets below the ``accuracy`` threshold.
+    """
+
+    def __init__(
+        self,
+        max_radial: int,
+        max_angular: int,
+        spline_cutoff: float,
+        accuracy: float,
+    ):
+        self.max_radial = max_radial
+        self.max_angular = max_angular
+        self.spline_cutoff = spline_cutoff
+        self.accuracy = accuracy
+
+    @abstractmethod
+    def _radial_integral(self, n: int, ell: int, positions: np.ndarray) -> np.ndarray:
+        """Method calculating the radial integral."""
+        ...
+
+    @property
+    def _center_contribution(self) -> Union[None, np.ndarray]:
+        r"""Contribution of the central atom required for LODE calculations.
+
+        The `center_contribution` is defined as :math:`c_n = \sqrt{4π} \int_0^\infty dr
+        r^2 R_n(r) g(r)` where :math:`g(r)` is the radially symmetric density function,
+        `R_n(r)` the radial basis function and :math:`n` the current radial channel.
+        This should be implemented in the child classes."""
+
+        return None
+
+    def _radial_integral_derivative(
+        self, n: int, ell: int, positions: np.ndarray
+    ) -> np.ndarray:
+        """Method calculating the derivatice of the radial integral."""
+
+        # The hardcoded displacement is a only a valid choice of the positions are in a
+        # reasonable range.
+        assert np.abs(positions).mean() >= 1.0
```

Comment:
Should we do a propper raise here? Probably it is better...

Reply:
would be nice yes =)